File: src/state/workspaces.ts
/* Workspace: nazwany zakres, w którym pracują agenci. JEDNA odpowiedź na pytanie „gdzie pracujemy".
 *
 * DECYZJA WŁAŚCICIELA, 2026-08-18. Do tego dnia folder pracy wybierało się **systemowym oknem,
 * przy każdym uruchomieniu biegu**: `launchRun` pytał o katalog, kiedy żadna karta nie była
 * otwarta. Zdanie właściciela po zobaczeniu tego okna brzmiało „mega chujnia" i było trafne —
 * wybór folderu jest decyzją o PROJEKCIE, podejmowaną raz, a nie czynnością powtarzaną przed
 * każdą pracą. Workspace jest więc nazwanym zakresem, wybieranym w bocznym menu, i to on mówi,
 * gdzie pracują agenci.
 *
 * CO WORKSPACE ZAKRESUJE, A CZEGO NIE. Wyłącznie **folder pracy i żywą sesję** (strumień, karty
 * biegów, stan biegu). Workflow, agenci, umiejętności i pamięć zostają GLOBALNE w `~/.loadout` —
 * powód stoi przy `src-tauri/src/commands/workspaces.rs` i jest nazwany: umiejętności piszą do
 * `~/.claude/skills` i `~/.agents/skills`, czyli do konfiguracji NARZĘDZI człowieka, a nie do
 * jego projektu. Zero migracji, zero drugiego katalogu.
 *
 * WORKSPACE ≠ KARTA, i to jest jedyna rzecz, którą trzeba tu czytać uważnie. Do 2026-08-18 ten
 * plik modelował KARTY FOLDERÓW (`WorkspaceTab`, `open`, `requestClose`) i słowo „workspace"
 * znaczyło w nim coś innego niż dziś. Ten kod nie zniknął — przeniósł się do
 * `src/state/run-tabs.ts` i jest re-eksportowany niżej, żeby przeprowadzka nie wywróciła pięciu
 * plików sekcji Run w trakcie. Rozdział znaczeń:
 *
 *   - **workspace** (ten plik) — nazwa + folder, wybierany w bocznym menu, TRWAŁY na dysku,
 *     jeden na całą aplikację jako „aktywny";
 *   - **karta** (`run-tabs.ts`) — bieg WEWNĄTRZ workspace, byt ulotny, żyje na ekranie Run.
 *
 * DYSK-PIERWSZY, WSZĘDZIE. `add`, `rename` i `remove` zmieniają stan DOPIERO po potwierdzeniu
 * z dysku i oddają `boolean`, żeby wołający zamknął formularz tylko wtedy, kiedy plik naprawdę
 * się zapisał. Odwrotna kolejność to defekt, który już raz w tym repo wystąpił: agent zniknięty
 * z listy przy NIEUDANYM usunięciu wracał po restarcie, bo okno uwierzyło sobie, a nie plikowi.
 *
 * ODMOWY IDĄ PRZEZ `why()`, NIGDY PRZEZ `instanceof Error`. Tauri odrzuca NAPISEM
 * (`src-tauri/src/ipc.rs` robi `.map_err(|e| e.to_string())`), więc warunek `error instanceof
 * Error` jest tu ZAWSZE fałszywy — stał w siedmiu miejscach i kasował każdą precyzyjną odmowę,
 * jaką Rust naprawdę pisze („The folder … is not there, so Loadout did not add it").
 */
import { create } from 'zustand';
import type { StoreApi, UseBoundStore } from 'zustand';

import { why } from '../ipc/why';
import { deleteWorkspace, listWorkspaces, saveWorkspace } from './workspaces-io';

/* SZEW MIGRACYJNY, 2026-08-18. Karty przeniosły się do `./run-tabs`, a te nazwy czyta dziś pięć
 * plików sekcji Run i trzy testy. Re-eksport nie jest drugą definicją (niezmiennik 13: ciało
 * stoi w jednym pliku) — jest przekierowaniem, które ma zniknąć w dniu, w którym te importy
 * wskażą `./run-tabs` wprost. Zapisane jako dług, nie jako rozwiązanie. */
export type {
  CancelRun,
  PendingClose,
  RunTabsState,
  WorkspaceTab,
  WorkspacesStore,
} from './run-tabs';
export { createWorkspacesStore } from './run-tabs';

/**
 * Jeden nazwany zakres pracy.
 *
 * `id` JEST ścieżką folderu (`WorkspaceWire` po stronie Rusta), i to nie jest oszczędność: jeden
 * folder = jeden workspace, więc ścieżka jest naturalnym kluczem i nie da się zapisać dwóch
 * wpisów o tym samym folderze przez pomyłkę. `folder` stoi osobno, bo okno nie ma prawa zakładać,
 * że klucz jest ścieżką — dzień, w którym klucz przestanie nią być, ma zmienić jeden plik.
 */
export interface Workspace {
  /** Klucz wpisu. Dziś równy `folder`. */
  readonly id: string;
  /** Nazwa nadana przez człowieka. To ona stoi w przełączniku bocznego menu. */
  readonly name: string;
  /** Folder pracy — jedyna rzecz, którą ten wpis naprawdę niesie. */
  readonly folder: string;
}

export interface WorkspacesState {
  /** Wszystkie zakresy, w kolejności zapisu na dysku. Pusta lista jest poprawnym stanem. */
  readonly all: readonly Workspace[];
  /** Zakres, w którym pracujemy; `null`, dopóki żadnego nie ma. */
  readonly activeId: string | null;
  /** Zdanie, którym dysk odmówił — dla człowieka, słowo w słowo od Rusta. `null`, kiedy nie odmówił. */
  readonly said: string | null;

  /** Czyta listę z dysku. Wołane raz, przy starcie okna. */
  load: () => Promise<void>;

  /** Dokłada zakres albo zmienia nazwę istniejącego (klucz to folder). `true` = dysk potwierdził. */
  add: (name: string, folder: string) => Promise<boolean>;

  /** Zmienia nazwę zakresu, który już jest na liście. `true` = dysk potwierdził. */
  rename: (id: string, name: string) => Promise<boolean>;

  /** Zdejmuje zakres z listy. Folderu nie dotyka. `true` = dysk potwierdził. */
  remove: (id: string) => Promise<boolean>;

  /**
   * Przełącza zakres. **Wyłącznie zmiana widoku** — dysku nie dotyka ani razu.
   *
   * Przełączenie nie ma prawa zgubić sesji i to jest wymóg twardy właściciela z 2026-08-18.
   * Ten magazyn go spełnia przez to, czego NIE robi: nie zatrzymuje biegu, nie odłącza pompy
   * linii i nie kasuje kart. Pompa należy do karty po stronie Rusta
   * (`src-tauri/src/workspace.rs`), nie do tego pola.
   */
  activate: (id: string) => void;

  /** Człowiek przeczytał odmowę. Nic poza zdaniem nie znika. */
  dismiss: () => void;
}

/**
 * Który zakres ma być aktywny, kiedy lista właśnie się zmieniła.
 *
 * JEDEN NIEZMIENNIK NA CAŁY MAGAZYN: `activeId` wskazuje na wpis, który ISTNIEJE, albo jest
 * `null`, kiedy nie ma żadnego. Bez tego zdania każda operacja musiałaby pamiętać własny
 * przypadek brzegowy, a `activeWorkspace()` oddawałoby `null` przy niepustej liście — czyli
 * ekran Run bez folderu przy trzech workspace'ach w menu.
 *
 * Pierwszy z listy, a nie `null`, kiedy życzenie przepadło: to samo zdanie obsługuje start okna
 * (po restarcie `activeId` nie istnieje, bo NIE JEST zapisywany na dysku — wybór widoku nie jest
 * stanem trwałym) i usunięcie aktywnego zakresu.
 */
function pick(all: readonly Workspace[], wanted: string | null): string | null {
  if (wanted !== null && all.some((one) => one.id === wanted)) return wanted;
  return all[0]?.id ?? null;
}

export const useWorkspaces: UseBoundStore<StoreApi<WorkspacesState>> = create<WorkspacesState>()(
  (set, get) => ({
    all: [],
    activeId: null,
    said: null,

    load: async () => {
      try {
        const all = await listWorkspaces();
        /* Pusta lista NIE jest błędem: na świeżej maszynie pliku nie ma i Rust oddaje `[]`.
         * Przełącznik pokaże wtedy zaproszenie do dodania pierwszego zakresu (DESIGN §6),
         * a nie zdanie o awarii. */
        set({ all, activeId: pick(all, get().activeId), said: null });
      } catch (error) {
        set({ said: why(error, 'Loadout could not read your list of workspaces.') });
      }
    },

    add: async (name, folder) => {
      try {
        /* DYSK PIERWSZY. Stan bierzemy z odpowiedzi, nie z argumentów: Rust przycina nazwę,
         * odmawia folderowi, którego nie ma, i przy DRUGIM zapisie tego samego folderu zmienia
         * nazwę zamiast dokładać wiersz. Lista złożona tutaj z `[...all, { name, folder }]`
         * pokazywałaby duplikat, którego w pliku nie ma. */
        const all = await saveWorkspace({ name, folder });
        /* Nowy zakres staje się aktywny, bo dodanie go JEST zdaniem „chcę tu pracować".
         * Wersja, która tylko dokłada wiersz, zostawia człowieka z listą i bez skutku. */
        set({ all, activeId: pick(all, folder), said: null });
        return true;
      } catch (error) {
        set({ said: why(error, 'Loadout could not add that workspace.') });
        return false;
      }
    },

    rename: async (id, name) => {
      const had = get().all.find((one) => one.id === id);
      /* Zmiana nazwy jedzie tą samą komendą co dodanie (klucz to folder), więc bez folderu nie
       * ma czego zapisać. Nieznany identyfikator to nie awaria dysku i nie wolno o nim mówić
       * zdaniem o dysku: lista zmieniła się pod ręką człowieka i to jest cała treść odmowy. */
      if (had === undefined) {
        set({ said: 'That workspace is no longer on the list, so Loadout did not rename it.' });
        return false;
      }
      try {
        const all = await saveWorkspace({ name, folder: had.folder });
        set({ all, activeId: pick(all, get().activeId), said: null });
        return true;
      } catch (error) {
        set({ said: why(error, 'Loadout could not rename that workspace.') });
        return false;
      }
    },

    remove: async (id) => {
      try {
        const all = await deleteWorkspace({ id });
        /* Zdjęcie AKTYWNEGO zakresu zwalnia życzenie — wtedy `pick` bierze pierwszy z listy albo
         * `null`. Zdjęcie zakresu w tle nie ma prawa przestawić widoku: to jest ta samowola, po
         * której „Remove" klika się z duszą na ramieniu, bo nie wiadomo, gdzie się wyląduje. */
        const wanted = get().activeId === id ? null : get().activeId;
        set({ all, activeId: pick(all, wanted), said: null });
        return true;
      } catch (error) {
        set({ said: why(error, 'Loadout could not remove that workspace.') });
        return false;
      }
    },

    activate: (id) => {
      /* Zakres, którego nie ma na liście, nie staje się aktywny — inaczej `activeWorkspace()`
       * oddawałoby wpis, którego nikt nie zapisał. Dysku ta metoda nie dotyka ani razu. */
      if (!get().all.some((one) => one.id === id)) return;
      set({ activeId: id });
    },

    dismiss: () => {
      set({ said: null });
    },
  }),
);

/**
 * Aktywny workspace albo `null`. JEDNA definicja „gdzie pracujemy" na całe repo.
 *
 * FUNKCJA MODUŁOWA, NIE HAK, i to jest cała jej konstrukcja: czytają ją rzeczy spoza drzewa
 * Reacta — kontrolka startu przed wysłaniem folderu do Rusta, kanał zdarzeń, kod uruchamiający
 * bieg. Hak dałby tę odpowiedź wyłącznie w czasie renderu, więc obok musiałaby powstać druga
 * droga do tej samej prawdy (niezmiennik 13).
 *
 * `activeWorkspace()?.folder` zastępuje `activeFolder()` z `src/sections/run/workspaces-store.ts`.
 * `null` znaczy „człowiek nie wskazał jeszcze zakresu", a odpowiedź na to pytanie należy do
 * Rusta (`AppState::project_for` bierze wtedy katalog, pod którym wstała aplikacja) — front,
 * który podstawiłby tu własną domyślną ścieżkę, byłby drugim miejscem, w którym mieszka
 * ta decyzja.
 */
export function activeWorkspace(): Workspace | null {
  const { all, activeId } = useWorkspaces.getState();
  if (activeId === null) return null;
  return all.find((one) => one.id === activeId) ?? null;
}

File: src-tauri/src/commands/workspaces.rs
//! Lista workspace'ów użytkownika: nazwa plus folder, w którym pracują agenci.
//!
//! **Ani jednego `use tauri::` i ani jednego `#[tauri::command]`** — jak w całym tym katalogu.
//! Skorupy stoją w `src/ipc.rs` i mają po dwie linie (niezmiennik 1).
//!
//! # Po co to istnieje
//!
//! 2026-08-18, decyzja Jakuba. Do tego dnia folder pracy wybierało się **systemowym oknem, przy
//! każdym uruchomieniu biegu**: `launchRun` pytał o katalog, jeśli żadna karta nie była otwarta.
//! Zdanie właściciela po zobaczeniu tego okna brzmiało „mega chujnia" i było trafne — wybór
//! folderu jest decyzją o PROJEKCIE, podejmowaną raz, a nie czynnością powtarzaną przed każdą
//! pracą. Workspace jest więc nazwanym zakresem, wybieranym w bocznym menu, i to on mówi,
//! gdzie pracują agenci.
//!
//! # Co workspace ZAKRESOWUJE, a czego nie
//!
//! Wyłącznie **folder pracy i żywą sesję** (strumień, karty biegów, stan biegu). Workflow,
//! agenci, umiejętności i pamięć zostają **globalne** w `~/.loadout` — to jest rozstrzygnięcie
//! z tego samego dnia i ma nazwany powód: umiejętności piszą do `~/.claude/skills`
//! i `~/.agents/skills` (`skills::DESTINATION_DIRS`), czyli do konfiguracji NARZĘDZI człowieka,
//! nie do jego projektu. Sekcja zakresowana per workspace i tak musiałaby się z tej reguły
//! wyłamać, a jedna sekcja łamiąca regułę kosztuje więcej niż brak reguły.
//!
//! Praktyczny skutek: agenta zapisanego raz widać w każdym workspace, a bieg zapisuje się pod
//! `<folder>/.loadout/runs/`, czyli w projekcie, którego dotyczy.
//!
//! # Czego tu ŚWIADOMIE nie ma, i to jest zgłoszenie
//!
//! Nie ma tu puli miejsc ani magazynu per folder. Oba mieszkają w `crate::workspace::Registry`
//! (546 linii, napisane 2026-08-16 i **do dziś bez ani jednego wołającego** — audyt nazwał to
//! wprost). Ten plik jest listą, którą widzi okno; tamten jest rejestrem, który powinien trzymać
//! `Store` na folder i JEDNĄ pulę miejsc na aplikację (`docs/ARCHITECTURE.md` §8, niezmiennik 11).
//! Wpięcie tamtego jest osobną zmianą: `AppState` trzyma dziś jeden `Store` otwarty na
//! `~/.loadout/loadout.db`, więc przeniesienie bazy pod folder projektu dotyka każdego biegu
//! i każdego kryterium, które go sądzi. Zgłoszone, nie zrobione po cichu.

use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

/// Nazwa pliku z listą. Jeden plik, nie katalog: to jest lista kilku wierszy, a nie biblioteka.
const FILE: &str = "workspaces.json";

/// Ile workspace'ów przyjmujemy. Sufit istnieje, bo lista jedzie w całości do bocznego menu,
/// a menu, przez które trzeba przewijać, przestaje być przełącznikiem.
pub const MOST_WORKSPACES: usize = 24;

/// Jeden workspace na drucie.
///
/// `id` jest **ścieżką folderu** i to nie jest oszczędność: jeden folder = jeden workspace, więc
/// ścieżka jest naturalnym kluczem i nie da się zapisać dwóch workspace'ów o tym samym folderze
/// przez pomyłkę. Ten sam wybór robi `crate::workspace::WorkspaceId`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceWire {
    /// Ścieżka folderu — klucz i jedyna rzecz, którą ten wpis naprawdę niesie.
    pub id: String,
    /// Nazwa nadana przez człowieka. To ona stoi w przełączniku.
    pub name: String,
    /// Folder pracy. Równy `id`; osobne pole, bo okno nie ma prawa zakładać, że klucz jest
    /// ścieżką — dzień, w którym klucz przestanie nią być, ma zmienić jeden plik, nie sześć.
    pub folder: String,
}

/// Dlaczego nie dało się zapisać albo przeczytać listy.
///
/// Każdy wariant jest osobnym zdaniem dla człowieka, bo każdy naprawia się inaczej (niezmiennik 14:
/// zero żargonu — „os error 2" nie mówi, co zrobić).
#[derive(Debug, thiserror::Error)]
pub enum WorkspaceError {
    /// Folder nie istnieje albo nie jest folderem.
    #[error(
        "The folder \"{folder}\" is not there, so Loadout did not add it. Pick a folder that \
         exists."
    )]
    NoSuchFolder {
        /// Ścieżka, którą podało okno.
        folder: String,
    },
    /// Ścieżka nie jest pełna.
    #[error(
        "Loadout needs the whole path to a folder, and \"{folder}\" is only part of one. Pick \
         the folder again."
    )]
    NotAWholePath {
        /// Ścieżka, którą podało okno.
        folder: String,
    },
    /// Nazwa jest pusta.
    #[error("Give this workspace a name first — that name is how you will pick it later.")]
    NoName,
    /// Lista jest pełna.
    #[error(
        "You already have {MOST_WORKSPACES} workspaces, which is as many as the switcher can \
         show. Remove one you no longer work in."
    )]
    TooMany,
    /// Pliku listy nie dało się przeczytać albo zapisać.
    #[error("Loadout could not save the list of workspaces: {0}")]
    Unwritable(#[source] io::Error),
    /// Plik listy jest uszkodzony.
    #[error(
        "The file that remembers your workspaces could not be read: {0}. Loadout left it alone \
         rather than overwrite it."
    )]
    Malformed(#[source] serde_json::Error),
}

/// Ścieżka pliku listy w bibliotece.
#[must_use]
pub fn list_path(home: &Path) -> PathBuf {
    home.join(FILE)
}

/// Workspace'y zapisane na dysku, w kolejności zapisu.
///
/// **Brak pliku to pusta lista, nie błąd.** Na świeżej maszynie tego pliku nie ma i to jest stan
/// normalny — dokładnie ta pomyłka (brakujący katalog traktowany jako awaria dysku) kończyła
/// każdy bieg zdaniem „No such file or directory (os error 2)".
///
/// Wpisy wskazujące na folder, którego już nie ma, **zostają na liście** i to jest wybór:
/// zniknięcie workspace'a z przełącznika, bo dysk zewnętrzny nie jest podłączony, wygląda jak
/// utrata pracy. Okno pokazuje je i mówi o nich; usuwa je człowiek.
pub fn list_workspaces_inner(home: &Path) -> Result<Vec<WorkspaceWire>, WorkspaceError> {
    let path = list_path(home);
    let text = match fs::read_to_string(&path) {
        Ok(text) => text,
        Err(error) if error.kind() == io::ErrorKind::NotFound => return Ok(Vec::new()),
        Err(error) => return Err(WorkspaceError::Unwritable(error)),
    };
    if text.trim().is_empty() {
        return Ok(Vec::new());
    }
    serde_json::from_str(&text).map_err(WorkspaceError::Malformed)
}

/// Dokłada workspace albo zmienia nazwę istniejącego, i oddaje listę po zapisie.
///
/// Oddaje **całą listę**, nie sam zapisany wpis, i to jest ta sama decyzja, co przy notatkach:
/// okno ma jedno źródło prawdy o liście i nie składa jej sobie z odpowiedzi na pojedyncze
/// zapisy. Lista złożona po stronie okna rozjeżdża się przy pierwszym zapisie, który częściowo
/// się nie udał.
///
/// Klucz to folder, więc drugi zapis tego samego folderu **zmienia nazwę**, a nie dokłada
/// drugiego wiersza. Dwa workspace'y nad jednym folderem to dwa magazyny nad jedną bazą
/// (niezmiennik 2) — czyli zakleszczenie, nie „czasem wolniej".
pub fn save_workspace_inner(
    home: &Path,
    name: &str,
    folder: &str,
) -> Result<Vec<WorkspaceWire>, WorkspaceError> {
    let name = name.trim();
    if name.is_empty() {
        return Err(WorkspaceError::NoName);
    }
    let path = PathBuf::from(folder);
    if !path.is_absolute() {
        return Err(WorkspaceError::NotAWholePath {
            folder: folder.to_owned(),
        });
    }
    // Sprawdzamy przy DOKŁADANIU, nie przy odczycie: folder, który zniknął po dodaniu, ma zostać
    // na liście (powód przy `list_workspaces_inner`), ale folder, którego nie ma w chwili
    // dodawania, jest pomyłką w oknie wyboru i mówimy o niej od razu.
    if !path.is_dir() {
        return Err(WorkspaceError::NoSuchFolder {
            folder: folder.to_owned(),
        });
    }

    let key = folder.to_owned();
    let mut all = list_workspaces_inner(home)?;
    if let Some(had) = all.iter_mut().find(|one| one.folder == key) {
        name.clone_into(&mut had.name);
    } else {
        if all.len() >= MOST_WORKSPACES {
            return Err(WorkspaceError::TooMany);
        }
        all.push(WorkspaceWire {
            id: key.clone(),
            name: name.to_owned(),
            folder: key,
        });
    }
    write(home, &all)?;
    Ok(all)
}

/// Zdejmuje workspace z listy i oddaje listę po zapisie.
///
/// **Folderu ani jego zawartości nie dotyka.** Usunięcie workspace'a jest zdjęciem zakresu
/// z przełącznika, a nie skasowaniem projektu — i to jest jedyna rzecz, którą ta komenda ma
/// prawo znaczyć. Katalog `.loadout/runs/` z historią biegów zostaje tam, gdzie był.
///
/// Nieznany identyfikator nie jest błędem: lista po tej operacji ma wyglądać tak samo,
/// niezależnie od tego, czy wpis był (idempotencja). Drugie kliknięcie „Remove" po odświeżeniu
/// listy w innym oknie nie ma prawa dać zdania o awarii.
pub fn delete_workspace_inner(home: &Path, id: &str) -> Result<Vec<WorkspaceWire>, WorkspaceError> {
    let mut all = list_workspaces_inner(home)?;
    all.retain(|one| one.id != id);
    write(home, &all)?;
    Ok(all)
}

/// Lista → plik, przez plik tymczasowy i `rename`.
///
/// `rename` w obrębie jednego katalogu jest atomowe: czytelnik widzi albo poprzednią listę
/// w całości, albo nową w całości, i nigdy zera bajtów w środku. Ta sama droga, którą zapisuje
/// się `run.json` (`commands::run::spill`) — bo ta sama klasa awarii kosztuje tu to samo:
/// przerwany zapis zabiera listę wszystkich projektów człowieka.
fn write(home: &Path, all: &[WorkspaceWire]) -> Result<(), WorkspaceError> {
    fs::create_dir_all(home).map_err(WorkspaceError::Unwritable)?;
    // Blad serializacji idzie w `Unwritable`, nie w `Malformed`: zdanie „nie dalo sie
    // PRZECZYTAC" wypowiedziane przy zapisie wysyla czlowieka szukac uszkodzonego pliku,
    // ktorego nie ma.
    let mut text = serde_json::to_string_pretty(all)
        .map_err(|error| WorkspaceError::Unwritable(io::Error::other(error)))?;
    // Znak nowej linii na końcu: bez niego każda zmiana ostatniego wiersza niesie w diffie
    // dodatkowe „\ No newline at end of file", a plik przestaje być zwykłym plikiem tekstowym.
    text.push('\n');
    let writing = list_path(home).with_extension("json.writing");
    fs::write(&writing, text).map_err(WorkspaceError::Unwritable)?;
    fs::rename(&writing, list_path(home)).map_err(WorkspaceError::Unwritable)
}
